File: exercism/elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map()
  def count(sentence) do
    to_lower_case = fn(w) -> String.downcase(w) end

    words = Regex.split(~r/[^\w-]|_/iu, sentence, trim: true)
    |> Enum.map(to_lower_case)

    merge_by_word = fn(word, acc) ->
      Map.merge(acc, %{word => 1}, fn _k, a, b -> a + b end)
    end

    Enum.reduce(words, %{}, merge_by_word)
  end
end

File: exercism/scala/word-count/src/main/scala/wordCount.scala
class Phrase(sentence: String) {
  def wordCount = {
    val words = ("""(\w|')+""".r).findAllIn(sentence)

    words.foldLeft(Map.empty[String, Int]) { case (acc, word) =>
      val count = acc.getOrElse(word.toLowerCase, 0)
      acc + (word.toLowerCase -> (count + 1))
    }
  }
}
